### ISSUE TITLE
Use historical data from before user started using the app in statistics (and elsewhere)

### ISSUE THREAD

**Issue by LefterisJP** [rotki/rotki#1379]
## Problem Definition

At the moment if a user gets premium and tries to use the statistics, he will see graphs with all the history of his holdings since starting to use the app. If a user wants to see history before that it's currently not possible. The statistics shown are from the data saved by Rotki. It saves snapshots of all of the user balances every 24 hours by default (you can edit the frequency in the settings).

Problem is you can't query kraken (or any exchange's) historical balances. The way to do it would be to perform true double entry accounting over all of the sources, replay all trades, all deposit withdrawals etc. and making some assumptions give some inaccurate estimates for balances. We don't do that as it's not yet possible. Perhaps we will devise a way to do that in the future.

## Task

Find a way to do this if any. Completing this task would be closely tied to true double entry accounting over everything in the app

**Comment by bengii** [comment]
same as problem statement in #1419

**Comment by LefterisJP** [comment]
No this is not the same. #1419 is talking about zapperfi, so I assume it means ethereum historical data. Historical data from an exchange are something completely different.

There is literally no way to query an exchange for its historical balance of a user. You can deduce it though by going through all trades/deposits/withdrawals, assuming your data is correct. This is what this issue is made for.

**Comment by bengii** [comment]
Thats clear, I did not know that

**Comment by emansipater** [comment]
Could this be why my tax reports always fail?  The lack of historical data?

**Comment by LefterisJP** [comment]
Hmm no, that would definitely be a different issue. No idea unless we debug with logs. You can send us logs in discord and/or make an issue.

A related issue we need to solve is: https://github.com/rotki/rotki/issues/1426

**Comment by Beanow** [comment]
The problem talks about quering exchanges, but does this issue also include historical wallet balances from the xpub/addresses?
Particularly the assets per account and fiat value at Jan 1st 0:00 would be very useful to find out for previous years.

**Comment by LefterisJP** [comment]
Hey @Beanow hello!

This is kind of a meta (epic) issue. Historical wallet balances would also be needed to create the whole picture. To address this we would need to create a lot of these meta issues, with historic balances being one of them.

January 1st historical balances + Netherlands as a bio points me to someone who wants to do tax accounting for the Netherlands where I have heard (from friends) that you are taxed based on balances on that day.

In that case we are trying to also switch towards a more [plugin/modular](https://github.com/rotki/rotki/issues/1874) architecture for most things, including accounting which would allow external people to write accounting plugins for their jurisdiction or use case.

Even though this is offtopic to the current issue I am putting it here just to let you know. We also got a discord server with dedicated development channel which you may join if you would like to help: https://discord.gg/aGCxHG7

**Comment by Beanow** [comment]
> January 1st historical balances + Netherlands as a bio points me to someone who wants to do tax accounting for the Netherlands where I have heard (from friends) that you are taxed based on balances on that day.

Indeed, for that particular tax you need a Euro estimate value of your assets at Jan 1st. And they're particular about which exchange rate to use. ([source](https://www.belastingdienst.nl/wps/wcm/connect/nl/werk-en-inkomen/content/cryptovaluta))

A module to make a specific report for this would be great. Though even just the balances per account at Jan 1st would save a lot of effort. I haven't found a good tools supporting a "balance at a point in time" feature from xpub/addresses. Having to work backwards from transaction histories is by far the most time consuming part.

So I ran into your scenario exactly. Subscribed, went to analytics with great hopes, but found no history. :disappointed:

For the long term, I'll consider contributing. But for now I just need a tool for this use-case or do it manually again.

**Comment by dnivi3** [comment]
>So I ran into your scenario exactly. Subscribed, went to analytics with great hopes, but found no history. 😞

Just wanted to chime in and say that I subscribed for the same reason as @Beanow in hopes that I would be able to see balances and worth at specific dates, but ran into the same problem.

**Comment by LefterisJP** [comment]
Hey @dnivi3 thanks a lot for getting a subscription. Every subscriber helps us towards the goal of keeping this project alive.

Some notes. Even when we implement this it's never going to be 100% accurate due to the amount of data required to be queried for this and the need for all data sources to be 100% accurate and able to provide all possible movements and prices at any point. It will be a best-effort implementation.

But this is something we will definitely implement. The current behaviour which is the most accurate one, but does not go back in time will still stay but will be togglable. So you could switch to between the current behaviour or the new behaviour in the stats page.

It will come, just needs time. We work 24/7. Stay on top of things by checking our discord: https://discord.gg/aGCxHG7

**Comment by mussonero** [comment]
@LefterisJP
Next time please _**Advertise**_ this feature is missing if the user just started using the app with a premium account. 

Reading the usage guide, it was not anywhere mentioned
https://rotki.readthedocs.io/en/latest/usage_guide.html#analytics

Thanks!

**Comment by LefterisJP** [comment]
Hey @mussonero thanks for your feedback. I added an entry in the docs for clarification here: https://github.com/rotki/rotki/pull/2691

To keep up to date with this and/or other things about the app make sure to join our discord: https://discord.gg/aGCxHG7

**Comment by defi-caratacus** [comment]
This would really improve user experience and add more value to the app

**Comment by sime** [comment]
I'd be willing to help test this.

Though I am another subscriber who signed up for a year on the idea historical rates would appear.

**Comment by alexvansande** [comment]
Hello issue from 2020. I come here to offer my support and hope you find a way to get to deployment. Maybe it would be easier if you focus instead of all dates, but rather a few points. Maybe once per year at the last day of the year, or at the last day of each quarter, with prices for only X largest coins, that would probably solve 90% of the issues.

**Comment by LefterisJP** [comment]
> Hello issue from 2020. I come here to offer my support and hope you find a way to get to deployment. Maybe it would be easier if you focus instead of all dates, but rather a few points. Maybe once per year at the last day of the year, or at the last day of each quarter, with prices for only X largest coins, that would probably solve 90% of the issues.

Makes no difference. Would require same amount of work. And this is a big one, non-trivial.

For historical balance on a specific date for a blockchain account that is different and assuming you have access to an arhive node should not be as hard to implement: https://github.com/rotki/rotki/issues/2227

**Comment by prettyirrelevant** [comment]
## Specification

1. Add a backend API that takes a single asset and a timestamp. Return amount and price in the user currency of that asset on that timestamp. Returned by processing events/ This should be probably on the per - asset view. Probably only available for premium.
**PR: https://github.com/rotki/rotki/pull/9223**

2. Add a backend API that takes a timestamp. Returns  a snapshot of all assets, amounts and prices in the user currency on that timestamp. Returned by processing events/ This should be probably be somewhere on accounting. So accounting - |
                                                                                                                                                                              | ->  Historical Snapshot
                                                                                                                                                                              | -> PnL Report
**PR: https://github.com/rotki/rotki/pull/9223**

3. Add a backend API that takes an asset and a time range. This will return the per asset amounts that will go to the asset amount view. We discussed that this should at first be the time points of all the events. So each time there is an event for the asset it's a time point. (In the future / Later) we could  add another parameter which would be the distance in which to aggregate events. So if two events are closer to each other than this distance you would aggregate them to one time point. The backend API endpoint will return the point as timestamp, amount. 
**PR: https://github.com/rotki/rotki/pull/9246**

4. The frontend will use the historical price query api endpoint to query then the price of the given asset at each point and calculate the value/price graph. If prices are missing it will show to the user a small warning and allow them to manually input them through a dialog. Gotta think how to do it when too many prices are missing. Also question to frontend here: How would you know for which days (outside of the existing events days) we have prices for?
**PR: https://github.com/rotki/rotki/pull/9265, https://github.com/rotki/rotki/pull/9271**

5. We then need to allow increase in points / resolution. That means query more prices for that asset on days between the start of asset history and now, for which we are missing prices. For each day we get a price, we gain a new point in the graph. And slowly the graph gets more data and becomes more useful. We need to query this one demand though. We should take over and rethink the way we do the historical price query of cryptocompare. Take that, generalize it and make it more easy to query and adapt to this use case.


6. Finally we have the net-worth graph. That is the hardest. Here we would absolutely need to do per-day aggregations. And query per-day prices for all assets. There should be an endpoint that does this aggregation per day and gets the total of each asset per day along with the manual price query of each in the user's selected currency for that day. This would return 1 point per day. With total value in user's currency per day. If any prices are missing from the historical manual oracle we would ignore them and note it down so that the user either adds them or they are queried later by the task that queries. Finally similarly to the per-asset view we would need to use the same logic to query more prices (and save them in the DB).
